### PR TITLE
implementing all iito compulsory functions. Just missing UpdateGiftInfo

### PR DESCRIPTION
**Pull request by Fabio752**
Just adding a few compulsory functions for IITO in a new file. Feel free to suggest anything on strategy or the actual implementation. It has been a quick change not tested enough, so I expect there are a few things that can be improved!

### CODE REVIEW

File: internal/clients/team4/iito.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+// GetGiftOffers allows clients to make offers in response to gift requests by other clients.
+// It can offer multiple partial gifts.
+// COMPULSORY, you need to implement this method. This placeholder implementation offers no gifts,
+// unless another team is critical.
+func (c *client) GetGiftOffers(receivedRequests shared.GiftRequestDict) shared.GiftOfferDict {
+	offers := shared.GiftOfferDict{}
+
+	ourResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
```
Comment (akowal3) [inline internal/clients/team4/iito.go]: in an ideal world, you could create a function to get resources, but it's fine

File: internal/clients/team4/iito.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	return shared.Resources(actualGift)
+}
+
+// SentGift is executed at the end of each turn and notifies clients that
+// their gift was successfully sent, along with the offer details.
+// COMPULSORY, you need to implement this method
+func (c *client) SentGift(sent shared.Resources, to shared.ClientID) {
+	// You can check your updated resources like this:
+	myResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
+	percentageDonated := float64(sent / (sent + myResources))
```
Comment (akowal3) [inline internal/clients/team4/iito.go]: div by 0 if send is 0?

File: internal/clients/team4/iito.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+}
+
+// ReceivedGift is executed at the end of each turn and notifies clients that
+// their gift was successfully received, along with the offer details.
+// COMPULSORY, you need to implement this method
+func (c *client) ReceivedGift(received shared.Resources, from shared.ClientID) {
+	// You can check your updated resources like this:
+	myResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
+	percentageDonated := float64(received) / float64(myResources)
```
Comment (akowal3) [inline internal/clients/team4/iito.go]: div by 0 if myresources is 0?

File: internal/clients/team4/iito.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+	// have a cap at 100 in case we are too selfless
+	wealthGoal := math.Max(selflessLevel * float64(ourResources), resourcesThreshold)
+	
+	// giftSize is the maximum we can afford giving away.
+	giftSize :=  float64(ourResources) - wealthGoal
+
+	// You can fetch the clients which are alive like this:
+	for team, status := range c.ServerReadHandle.GetGameState().ClientLifeStatuses {
+		teamTrust := c.trustMatrix.GetClientTrust(team)
```
Comment (akowal3) [inline internal/clients/team4/iito.go]: you can also use c.getTrust(team)

File: internal/clients/team4/iito.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+	currentCollab := c.internalParam.collaboration
+	maxCollabBump := math.Min(currentCollab + bumpThreshold, 1.0)
+
+	currentAgentTrust := c.trustMatrix.GetClientTrust(from)
+	maxTrustBump := math.Min(currentAgentTrust + bumpThreshold, 1.0)
+	// Bump up collaboration and agent trust.
+	// We use a maximum bump to cap at 1 and reduce the volatility.
+	c.internalParam.collaboration = math.Min(currentCollab + percentageDonated, maxCollabBump)
+	c.trustMatrix.SetClientTrust(from, math.Min(currentAgentTrust + percentageDonated, maxTrustBump))
```
Comment (akowal3) [inline internal/clients/team4/iito.go]: this discards our previously built trust if maxTrustBump is reached. Is that expected?

**Review by akowal3 — CHANGES_REQUESTED** [review]
see comments

File: internal/clients/team4/iito.go (hunk 5 above)
Comment (Fabio752) [reply internal/clients/team4/iito.go]: This is resolved based on our chat.

File: internal/clients/team4/utilities.go [hunk 6]
```diff
@@ -124,3 +124,6 @@ func boolToFloat(input bool) float64 {
 	return 0
 }
 
+func (c *client) getOurResources() shared.Resources {
+	return c.ServerReadHandle.GetGameState().ClientInfo.Resources
+}
```
Comment (tobhil98) [inline internal/clients/team4/utilities.go]: Add if statement around like the other utils functions to not break tests

File: internal/clients/team4/iito.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	
+
+}
+
+// ReceivedGift is executed at the end of each turn and notifies clients that
+// their gift was successfully received, along with the offer details.
+// COMPULSORY, you need to implement this method
+func (c *client) ReceivedGift(received shared.Resources, from shared.ClientID) {
+	// You can check your updated resources like this:
+	myResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
```
Comment (tobhil98) [inline internal/clients/team4/iito.go]: Use getOurResources() here

File: internal/clients/team4/iito.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	actualGift := math.Min(float64(giftOffer), giftSize)
+	return shared.Resources(actualGift)
+}
+
+// SentGift is executed at the end of each turn and notifies clients that
+// their gift was successfully sent, along with the offer details.
+// COMPULSORY, you need to implement this method
+func (c *client) SentGift(sent shared.Resources, to shared.ClientID) {
+	// You can check your updated resources like this:
+	myResources := c.ServerReadHandle.GetGameState().ClientInfo.Resources
```
Comment (tobhil98) [inline internal/clients/team4/iito.go]: Use getOurResources here

**Pull request merged**